src/interpreter.rs

```diff
@@ -18,6 +19,95 @@ impl Execute for Expression {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+pub enum OperandItem<'a> {
+    Val(bool),
+    Predicate(&'a Predicate),
+}
+fn evaluate_operand_item(item: OperandItem, ctx: &mut Context, m: &mut Match) -> bool {
+    match item {
+        OperandItem::Val(b) => b,
+        OperandItem::Predicate(p) => p.execute(ctx, m),
+    }
+}
+
+impl Execute for Lir {
+    fn execute(&self, ctx: &mut Context, m: &mut Match) -> bool {
+        //operand stack
+        let mut top: usize = 0;
+        let mut operand_stack: [OperandItem; 2] = [OperandItem::Val(false); 2];
+
+        for code in &self.codes {
+            match code {
+                LirCode::LogicalOperator(op) => {
+                    match op {
+                        LirLogicalOperators::And => {
+                            let left = evaluate_operand_item(operand_stack[top - 2], ctx, m);
+                            if !left {
+                                // short circuit
+                                // stack pop
+                                top -= 2;
+
+                                //stack push
+                                operand_stack[top] = OperandItem::Val(false);
+                                top += 1;
+                            } else {
+                                let right = evaluate_operand_item(operand_stack[top - 1], ctx, m);
+
+                                //stack pop
+                                top -= 2;
+                                //stack push
+                                operand_stack[top] = OperandItem::Val(right);
+                                top += 1;
+                            }
+                        }
+                        LirLogicalOperators::Or => {
+                            let left = evaluate_operand_item(operand_stack[top - 2], ctx, m);
+                            if left {
+                                // short circuit
+                                // stack pop
+                                top -= 2;
+
+                                //stack push
+                                operand_stack[top] = OperandItem::Val(true);
+                                top += 1;
+                            } else {
+                                let right = evaluate_operand_item(operand_stack[top - 1], ctx, m);
```

Comment:
Actually, it can be simplified as:

```rust
operand_stack[0] = evaluate_operand_item(operand_stack[top - 2], ctx, m) or evaluate_operand_item(operand_stack[top - 1], ctx, m);
top = 1;
```

Which is more intuitive and straightforward, and also shorter.

Reply:
This also could apply to the And Operator. Using the short-circuit provided by language first will be more straightforward.